cpu/xmc1000/periph/uart.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2015 Sebastian Sontberg <sebastian@sontberg.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_xmc1000
+ * @{
+ *
+ * @file
+ * @brief       Low-level UART driver implementation
+ *
+ * @author      Sebastian Sontberg <sebastian@sontberg.de>
+ *
+ * @}
+ */
+
+#include "periph/gating.h"
+#include "periph/gpio.h"
+#include "periph/uart.h"
+#include "periph_conf.h"
+#include "sched.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/* guard file in case no UART device was specified */
+#if UART_NUMOF
+
+/**
+ * @brief Allocate memory to store the callback functions.
+ */
+static uart_isr_ctx_t uart_ctx[UART_NUMOF];
+
+static inline USIC_CH_TypeDef *_usic(uart_t uart)
+{
+    return uart_dev[uart].usic;
+}
+
+int uart_init(uart_t uart, uint32_t baudrate,
+              uart_rx_cb_t rx_cb, uart_tx_cb_t tx_cb, void *arg)
+{
+    USIC_CH_TypeDef *usic = _usic(uart);
+    gpio_t uart_tx = uart_dev[uart].tx;
+    gpio_t uart_rx = uart_dev[uart].rx;
+
+    /* clear gating, enable clock */
+    GATING_CLEAR(USIC0);
+
+    /* enable the USIC module */
+    usic->KSCFG |= USIC_CH_KSCFG_MODEN_Msk | USIC_CH_KSCFG_BPMODEN_Msk;
+
+    /* TODO: Support more rates */
+    if ((115200 % baudrate) != 0 || (baudrate < 28800)) {
+        return -2;
+    }
+
+    /* "The fractional divider register FDR allows the generation of the
+       internal frequency f_FD, that is derived from the system clock f_PB." */
+    /* uint32_t fdr_step = 590; */
```

Comment:
Don't bother with the uart driver please. It will share the majority of it's code with the SPI one, which already looks a lot better although it's not working yet ;)